src/App/Fossa/LicenseScanner.hs

```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
+    (_, True) -> do
+      case skippedDeps of
+        [] -> logDebug "All vendored dependencies are un-scanned and require license scanning"
+        _ -> logDebug . pretty $ "Some vendored dependencies have already been scanned by FOSSA. Skipping these vendored dependencies: " <> show skippedDeps
+
+  -- At this point, we have a good list of deps, so go for it.
+  maybeScannedArchives <- traverse (scanAndUploadVendoredDep baseDir) needScanningDeps
+
+  -- We need to include both scanned and skipped archives in this list so that they all get included in the build in FOSSA
+  let skippedArchives = map forceVendoredToArchive skippedDeps
+  archives <- fromMaybe NoSuccessfulScans $ NE.nonEmpty $ (catMaybes maybeScannedArchives) <> skippedArchives
+
+  -- finalizeLicenseScan takes archives without Organization information. This orgID is appended when creating the build on the backend.
+  -- We don't care about the response here because if the build has already been queued, we get a 401 response.
+  finalizeLicenseScan $ ArchiveComponents $ NE.toList archives
+
   pure $ NE.map arcToLocator (archivesWithOrganization orgId archives)
   where
     archivesWithOrganization :: OrgId -> NonEmpty Archive -> NonEmpty Archive
-    archivesWithOrganization orgId = NE.map $ includeOrgId orgId
+    archivesWithOrganization org = NE.map $ includeOrgId org
 
     includeOrgId :: OrgId -> Archive -> Archive
-    includeOrgId orgId arc = arc{archiveName = showT orgId <> "/" <> archiveName arc}
+    includeOrgId org arc = arc{archiveName = showT org <> "/" <> archiveName arc}
+
+findDepsThatNeedScanning ::
+  ( Has (Lift IO) sig m
+  , Has FossaApiClient sig m
+  ) =>
+  NonEmpty VendoredDependency ->
+  OrgId ->
+  m ([VendoredDependency], [VendoredDependency])
+findDepsThatNeedScanning vdeps orgId = do
+  revisionInfos <- getRevisionInfo vdeps
+  pure $ NE.partition (shouldScanRevision revisionInfos orgId) vdeps
+
+ensureVendoredDepVersion ::
+  (Has (Lift IO) sig m) =>
+  Path Abs Dir ->
+  VendoredDependency ->
+  m VendoredDependency
+ensureVendoredDepVersion baseDir vdep = do
+  depVersion <- case vendoredVersion vdep of
+    Nothing -> sendIO $ withSystemTempDir "fossa-temp" (calculateVendoredHash baseDir (vendoredPath vdep))
+    Just version -> pure version
+  pure vdep{vendoredVersion = Just depVersion}
+
+shouldScanRevision :: [RevisionInfo] -> OrgId -> VendoredDependency -> Bool
+shouldScanRevision revisionInfos orgId vdep = all (locatorMatches vdep orgId) revisionInfos
+  where
+    locatorMatches :: VendoredDependency -> OrgId -> RevisionInfo -> Bool
+    locatorMatches VendoredDependency{..} org RevisionInfo{..} =
+      locatorUrl /= revisionInfoLocator || not revisionInfoResolved
+      where
+        locator = Locator{locatorFetcher = "archive", locatorProject = vendoredName, locatorRevision = vendoredVersion}
+        locatorUrl = renderLocatorUrl org locator
```

Comment:
If I'm reading this correctly, we will not rescan a vendored dependency if:
- The version in `fossa-deps.yaml` has not changed, and
- A file on disk inside the directory has changed (I think we're only hashing if there is no version provided).

If I'm correct, I think that's a little unintuitive, but I'm okay with it if that's intentional. Will you please update the documentation and add a doc comment to this function making it clear this is intended functionality?

Reply:
You're right, this is not at all intuitive.

This logic was adopted from how archive uploads work.

The reasoning is that even if we did run a license scan on the dependency, an already analyzed revision exists on core already, so the build would not get queued and the data we send up would just be thrown away.

I'll add a comment to explain this